### DISCUSSION TITLE
HELP! Conditionally source a tool if it exists otherwise do nothing

### DISCUSSION THREAD

**Discussion by Or1g3n** [nushell/nushell#14482] [Q&A] [answered]
I am trying to do what would be dead simple in bash and have spent a couple hours at this point in pure frustration.

In this time I have learned that Nu has a parse then evaluation structure and this is by design which means what would work in bash will not in nu.

My use case is to only source zoxide if it's installed. 

Since ``` ( which zoxide | is-not-empty ) ``` is unsupported, the best alternative was ``` ( ~/.zoxide.nu | path exists ) ```. The problem is even though I have effectively added a condition that is supported, it is practically useless. The reason is I cannot have source inside the true condition because then the source is bound inside the block and doesn't actually do anything....

So after much searching found this pattern from the docs and after reading several issues or reddit posts:
```
const some_path = $nu.default-config-dir
source $"($some_path)/config.nu"
```
The problem is, again, at least in my use case, even though this works, its practically useless. Here is how I would implement what I need using this pattern:
```
const zi_path = ( if  ( ~/.zoxide.nu | path exists ) { '~/.zoxide.nu' } )
source zi_path
```
This is useless because even though it's syntactically sound, it's quite stupid because if zoxide path doesn't exist then the source will fail which defeats the whole purpose of the if condition.

If only there was a way to run source only if a condition is true like:
```
if ( ~/.zoxide.nu | path exists ) {
    source '~/.zoxide.nu' 
} else { 
    print 'zoxide.nu not found'
} 
```

Can someone who is well-versed in nushell please help me out?! Is what I am trying to do not possible?

**Comment by ghost** [comment]
nushell `source`, `use`, `overlay`, etc only work if the imported file exists at parse time (happens before execution time).

for nushell configs the workaround has for a long time been to generate nushell code as text in `env.nu` based on whatever the environment is, then write it into a temporary file, and then `source` that file in `config.nu`.  
this works since `config.nu` gets parsed after `env.nu` gets executed.

in your case you could take a shortcut and just say `if (not ("~/.zoxide.nu" | path exists)) { "" | save -r ~/zoxide.nu }` in your `env.nu` and then just `source ~/.zoxide.nu` in your `config.nu` without another file.

a completely different solution: i use a package manager for my nu config and have a per-device package-list. this way i don't have to write anything twice and don't have these issues (and it also fixes a bunch of other ones).  
im currently using [my own one](https://github.com/Jan9103/numng), but there is also [one by the nushell team](https://github.com/nushell/nupm)

> source zi_path

you forgot a `$` here.  
also: the variable could be null, which wouldn't work.  
and nushell only allows constants in source statements.

> if ( \~/.zoxide.nu | path exists ) { source '\~/.zoxide.nu'

nushell has very strict scoping. if this would work zoxide would only exist within that if statement.  
additionally it does not work since nu checks all sourced files at parse time, making the check useless.

**Reply by Or1g3n** [reply]
@Jan9103 
Thanks a lot for this reponse. I like the package manager idea, will spend next weekend figuring that out as it seems to be the ideal way to  go.

**Comment by fdncred** [comment, marked as answer]
Another thought is to try and use our `$nu.vendor-autoload-dirs` folders. The files in these folders are sourced one at a time. So, you could write a file to one of those locations and they're automatically sourced every time nushell loads.

For instance, here's mine with a file in each of the folders (except 1).
```nushell
❯ $nu.vendor-autoload-dirs
╭───┬────────────────────────────────────────────────────────────────────╮
│ 0 │ /Library/Application Support/nushell/vendor/autoload               │
│ 1 │ /usr/share/nushell/vendor/autoload                                 │
│ 2 │ /usr/local/share/nushell/vendor/autoload                           │
│ 3 │ /Users/fdncred/.local/share/nushell/vendor/autoload                │
│ 4 │ /Users/fdncred/Library/Application Support/nushell/vendor/autoload │
╰───┴────────────────────────────────────────────────────────────────────╯
```
exit and restart nushell
```nushell
❯ nu
Hello from env.nu!
Hello from config.nu!
Hello from defs.nu!
Hello from /Library/Application Support/nushell/vendor/autoload/01_autoload.nu
Hello from /usr/local/share/nushell/vendor/autoload/02_autoload.nu
Hello from /Users/fdncred/.local/share/nushell/vendor/autoload/03_autoload.nu
Hello from /Users/fdncred/Library/Application Support/nushell/vendor/autoload/04_autoload.nu
     __  ,
 .--()°'.' Welcome to Nushell,
'|, . ,'   based on the nu language,
 !_-(_\    where all data is structured!
```

**Reply by Or1g3n** [reply]
@Jan9103 @fdncred 

Really appreciate both of your responses.

Here is what I did trying out Darren's idea.

It is a bit hacky but it worked. I removed the source code from my config.nu then added the following to my env.nu:
```
# Zoxide
if (which zoxide | is-not-empty) { zoxide init nushell | save -f C:/Users/chris/AppData/Roaming/nushell/vendor/autoload/zoxide.nu }
if ( 'C:/Users/chris/AppData/Roaming/nushell/vendor/autoload/zoxide.nu' | path exists ) {
    'alias cd = z' | save -a ~/AppData/Roaming/nushell/vendor/autoload/zoxide.nu
}
```
One of the challenges i ran into was adding the alias. Typically i would have an aliases.nu that also had a condition for zoxide and added the associated alias. However, first i ran into an order of operations issue bc I had to ensure that the alias file was sourced last, then i ran into the same issue with scoping where the alias could not be inside of an if block. So i trashed the aliases.nu idea and added the alias directly in the zoxide code block in mt env.nu.
I could make this better by making the hard coded paths more dyanmic but it works for now.

@fdncred, is my approach close to what you intended? Also, as a maintainer, do you foresee this type of stuff being easier in the future?

**Reply by fdncred** [reply]
ya, that's what i was talking about. it may get easier in the future, not sure. we're pretty strict on scoping but i could see having an `if --env` that helps similar to `def --env`.
